lib/SocketManager.ts

```diff
@@ -9,41 +9,59 @@ const logger = module('Socket')
 
 // FIXME: this constants are duplicated on /src/plugins/socket.js. When converting this to ES6 module that can be removed
 // events from server ---> client
-export const socketEvents = {
-	init: 'INIT', // automatically sent when a new client connects to the socket
-	controller: 'CONTROLLER_CMD', // controller status updates
-	connected: 'CONNECTED', // socket status
-	nodeAdded: 'NODE_ADDED',
-	nodeRemoved: 'NODE_REMOVED',
-	nodeUpdated: 'NODE_UPDATED',
-	valueUpdated: 'VALUE_UPDATED',
-	valueRemoved: 'VALUE_REMOVED',
-	healProgress: 'HEAL_PROGRESS',
-	info: 'INFO',
-	api: 'API_RETURN', // api results
-	debug: 'DEBUG',
+export enum socketEvents {
+	init = 'INIT', // automatically sent when a new client connects to the socket
+	controller = 'CONTROLLER_CMD', // controller status updates
+	connected = 'CONNECTED', // socket status
+	nodeAdded = 'NODE_ADDED',
+	nodeRemoved = 'NODE_REMOVED',
+	nodeUpdated = 'NODE_UPDATED',
+	valueUpdated = 'VALUE_UPDATED',
+	valueRemoved = 'VALUE_REMOVED',
+	healProgress = 'HEAL_PROGRESS',
+	info = 'INFO',
+	api = 'API_RETURN', // api results
+	debug = 'DEBUG',
 }
 
 // events from client ---> server
-export const inboundEvents = {
-	init: 'INITED', // get all nodes
-	zwave: 'ZWAVE_API', // call a zwave api
-	hass: 'HASS_API', // call an hass api
-	mqtt: 'MQTT_API', // call an mqtt api
+export enum inboundEvents {
+	init = 'INITED', // get all nodes
+	zwave = 'ZWAVE_API', // call a zwave api
+	hass = 'HASS_API', // call an hass api
+	mqtt = 'MQTT_API', // call an mqtt api
+}
+
+declare interface SocketManager {
+	on(event: inboundEvents.init, listener: (socket: Socket) => void): this
+
+	on(
+		event: inboundEvents.zwave,
+		listener: (socket: Socket, data) => void
+	): this
+
+	on(
+		event: inboundEvents.hass,
+		listener: (socket: Socket, data) => void
+	): this
+
+	on(
+		event: inboundEvents.mqtt,
+		listener: (socket: Socket, data) => void
+	): this
 }
```

Comment:
I suggest you take a look at https://github.com/zwave-js/node-zwave-js/blob/4d9d317edf5192bec654d229fb8bc86c1966df29/packages/shared/src/EventEmitter.ts#L4 and https://github.com/zwave-js/node-zwave-js/blob/4d9d317edf5192bec654d229fb8bc86c1966df29/packages/zwave-js/src/lib/driver/Driver.ts#L326-L340

Reply:
That's black magic 🤣